### ISSUE TITLE
Decoupling tensorflow from keras in stubs

### ISSUE THREAD

**Issue by Avasam** [python/typeshed#11714]
Whilst working on https://github.com/python/typeshed/pull/11696, I was having issues referencing the source code. After reading the doc and a bit of investigation, I realized why:

The runtime of https://github.com/python/typeshed/tree/main/stubs/tensorflow/tensorflow/keras actually are just references to the keras source code at https://github.com/keras-team/keras/tree/master/keras (`import keras`), and is generated / injected in `import tensorflow.keras` by https://github.com/keras-team/keras/blob/6454a4888a494c20ab0ea1dc6912cbcc13c5f940/pip_build.py#L62

![image](https://github.com/python/typeshed/assets/1350584/6cf61401-ef20-4b7f-bd35-d3165e133323)

In other words, keras "pollutes" the tensorflow namespace! I don't know if the type stubs specifications allow two distributions to write to the same stub-only package, which would allow us to perfectly reflect what's truly happening. Second best thing would be to have stubs for `keras` and keep `tensorflow.keras` in the tensorflow stubs.

**Comment by hmc-cs-mdrissi** [comment]
The actual way source code looks for tensorflow is not goal of stubs. Tensorflow does a lot of dynamic import behaviors and re-exports symbols using decorators (tf_export) in other modules. The exact placement is unstable and changes regularly in very breaking ways. Keras source code has changed rather drastically in past several versions with folders being added to wheel that may not even exist in GitHub repository either.

Instead the goal of the stubs is to model the documentation and public api surface. tf.keras.optimizers.Optimizer is publicly documented as the type and works at runtime. Source may have it in keras package as keras.optimizers.Optimizer. Or maybe it’s keras.src.optimizers.Optimizer in the wheel but not repository. And tensorflow/keras were separate sources several years ago, got merged into one repo, and then got separated into two repos again years later. I do not want stubs to capture organizational implementation choices not captured by documentation/stability expectations of the library.

If you would like to add keras stubs folder for publicly documented members with keras root and then re-export them in tensorflow other publicly documented aliases I would be comfortable and supportive of that. I push back strongly on trying to mimic source code implementation though and any details not part of documentation.

**Comment by Avasam** [comment]
Thank you, that's exactly the kind of comments and historical insight I was looking for.

> If you would like to add keras stubs folder for publicly documented members with keras root and then re-export them in tensorflow other publicly documented aliases I would be comfortable and supportive of that.

I'd like to do that yes. Given how tightly coupled keras seems to be to tensorflow, I wouldn't mind if the package was exposed as a top-level under tensorflow stubs. (With only the types re-exported by tensorflow)

Since the source code actually exists on disk (not a c-extension or dynamically generated module), I do believe there's value added for both devs using the stubs and typeshed contributors, to be able to access the source from within their working environment / IDE

**Comment by hmc-cs-mdrissi** [comment]
On Keras/Tensorflow coupling, originally Keras was high level wrapper of tensorflow and theano. Theano is an older ml library before tensorflow. Tensorflow was mostly upgrade over theano so several years ago theano stop being maintained and recommended its users move to other ml libraries. After this for years Keras only supported tensorflow.

Recently (last couple months), Keras 3 came out and now Keras is trying to be high level wrapper of multiple major ml libraries again. This time Keras supports tensorflow, Jax, and PyTorch.

My own personal experience is mostly focused on tensorflow though. I have moderate experience doing projects in PyTorch and have never used jax.

**Comment by hmc-cs-mdrissi** [comment]
The current status is pip install types-tensorflow. That will give you all that is available today.

There are many apis not covered today and they continue to gradually be added at pace of open source. Tensorflow apis tend to be better than keras specific ones and I know keras 3 is major change from keras 2. The codebase I work on is still using Keras 2 so apis new to Keras 3 are not ones I’ll likely look at soon.
